server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java

```diff
@@ -0,0 +1,174 @@
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.queue.jms;
+
+import static org.apache.james.queue.api.Mails.defaultMail;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import java.util.stream.IntStream;
+
+import javax.mail.MessagingException;
+
+import org.apache.james.metrics.api.Gauge;
+import org.apache.james.metrics.api.GaugeRegistry;
+import org.apache.james.metrics.api.Metric;
+import org.apache.james.queue.api.MailQueue;
+import org.apache.james.queue.api.MailQueueContract;
+import org.apache.mailet.base.test.FakeMail;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+
+import com.github.fge.lambdas.Throwing;
+
+@ExtendWith(JMSMailQueueMetricExtension.class)
+public interface JMSMailQueueMetricContract extends MailQueueContract {
+
+    default FakeMail fakeMail() throws MessagingException {
+        return defaultMail()
+            .name("name1")
+            .build();
+    }
+
+    default void enQueueMail(JMSMailQueueMetricExtension.JMSMailQueueMetricTestSystem testSystem, Integer times) throws Exception {
+        IntStream
+            .rangeClosed(1, times)
+            .forEach(Throwing.intConsumer(time -> testSystem.getMockMailQueue().enQueue(fakeMail())));
+    }
+
+    default void deQueueMail(JMSMailQueueMetricExtension.JMSMailQueueMetricTestSystem testSystem, Integer times) throws Exception {
+        IntStream
+            .rangeClosed(1, times)
+            .forEach(Throwing.intConsumer(time -> testSystem.getMockMailQueue().deQueue().done(true)));
+    }
+
+    @Test
+    default void enqueueShouldIncreaseMetric(JMSMailQueueMetricExtension.JMSMailQueueMetricTestSystem testSystem) throws Exception {
+        Metric enqueuedMailsMetric = testSystem.getMockEnqueuedMailsMetric();
+
+        enQueueMail(testSystem, 2);
+
+        verify(enqueuedMailsMetric, times(2)).increment();
+        verifyNoMoreInteractions(enqueuedMailsMetric);
+    }
+
+    @Test
+    default void enqueueShouldRegisterGauge(JMSMailQueueMetricExtension.JMSMailQueueMetricTestSystem testSystem) throws Exception {
```

Comment:
No. Creating the mailQueue should register the gauge.

And when I call the gauge, MailQueue::getSize needs to be called.

Note that implementing this test needs to SPY mailqueue and use argument collection to retrieve the gauge

Reply:
There is a little tricky of `spy()` method, it cannot verify the `spider` method which will call a method reference inside. I did an alternative assertion in my `JMSMailQueueMetricContract.constructorShouldRegisterGetQueueSizeGauge`